rust/datafusion/src/execution/context.rs

```diff
@@ -417,7 +445,16 @@ impl ExecutionConfig {
         mut self,
         physical_planner: Arc<dyn PhysicalPlanner>,
     ) -> Self {
-        self.physical_planner = Some(physical_planner);
+        self.physical_planner = physical_planner;
+        self
+    }
+
+    /// Optional source of additional optimization passes
+    pub fn with_optimizer_rule_source(
```

Comment:
Going in the same lines as my comment about the planning, one idea is to make `optimizer_rule_source` be a vector of rules, and instantiate `ExecutionConfig` with a `vec` of those rules, instead of having an extra `OptimizerRuleSource` to hold those rules.

The rules themselves are `Arc<dyn OptimizerRule>`, and users write them by the trait.

Reply:
I would like to implement the pattern of passing in  `Arc<dyn OptimizerRule>` directly (in fact I tried to implement it initially). The issue I found is that `ExecutionContext` is read only (immutable) during planning, but the signature of `OptimizerRule` requires `&mut self` (aka it can change the `OptimizerRule` state).

I think Rust is actually saving us from ourselves, because as discussed in  https://issues.apache.org/jira/browse/ARROW-9888 the ExecutionContext can be shared between threads (and thus allowing one thread to mutate the `OptimizerRule`s could cause problems) 

What do you think about making `optimizer_rules` `Vec<dyn OptimizerRule + Clone>` and then planning with the result of `clone()`ing ? 

Actually, the more I think about this the more I like the `Clone` idea...

Reply:
I like with `Clone` also.

Another idea (not for this PR) is to make `OptimizerRule` be a function only, i.e. replace the trait by `Fn(Plan) -> Result<Plan>`. `TypeCoercer` lost its state in one of my recent PRs. In the end, the way we have been using this trait is:

1. Initialize the optimizer without arguments
2. Run the optimizer with a single argument (logicalplan)
3. remove the optimizer

If we really need to offer the optimizer some more information, we could just pass the executionContextState as a second argument of `optimize`.

This way, we do not really need to worry about memory management, which we really do not need in practice, as optimizations are stateless unless we need some introspection as to what it did, in which case logging seems adequate.

Reply:
@jorgecarleitao got it -- I will perhaps try with Clone later. There is  a style of writing optimizers that actually does put state on the optimizer object itself -- another style, which I think you used for predicate pushdown, is to pass state as explicit arguments.

The challenge with using function arguments is that I can't think of any way to write generic functions like `optimize_children` in that style: https://github.com/apache/arrow/pull/8097/files#diff-4a428d98d7a9b116e888c2abd108fbedR32-R43.

I'll think about it this evening and maybe something will pop into my head

Reply:
As it turns out that @andygrove 's suggestion for a single planner trait has removed the need for this trait entirely 🎉